src/staking/manager.ts

```diff
@@ -0,0 +1,660 @@
+import { getBbnParamByBtcHeight, getBbnParamByVersion } from "../utils/staking/param";
+import { reverseBuffer, uint8ArrayToHex } from "../utils";
+
+export interface BtcProvider {
+  // Sign a PSBT
+  signPsbt(psbtHex: string): Promise<string>;
+  // Sign a message using the ECDSA type
+  signMessage: (message: string, type: "ecdsa") => Promise<string>;
+  // Get the staker info
+  getStakerInfo(): Promise<StakerInfo>;
+  // Get the available UTXOs for staking
+  getUTXOs(): Promise<UTXO[]>;
+}
+
+export interface BbnProvider {
+  // Get the Babylon BTC tip height. Note the there might be a delay between
+  // the BTC tip height and the Babylon BTC tip height. You should not use the
+  // BTC tip height to determine the BBN param.
+  getBabylonBtcTipHeight(): Promise<number>;
+  // Get the Babylon Chain address
+  getBbnAddress(): Promise<string>;
+  // Sign a babylon transaction.
+  signTransaction: <T extends object>(msg: {
+    typeUrl: string;
+    value: T;
+  }) => Promise<Uint8Array>
+}
+
+interface StakingInputs {
+  finalityProviderPkNoCoordHex: string;
+  stakingAmountSat: number;
+  stakingTimelock: number;
+}
+
+interface InclusionProof {
+  pos: number;
+  merkle: string[];
+  blockHash: string;
+}
+
+export class BabylonBtcStakingManager {
+  private stakingParams: VersionedStakingParams[];
+  private btcProvider: BtcProvider;
+  private network: networks.Network;
+  private bbnProvider: BbnProvider;
+
+  constructor(
+    network: networks.Network,
+    stakingParams: VersionedStakingParams[],
+    btcProvider: BtcProvider,
+    bbnProvider: BbnProvider,
+  ) {
+    this.network = network;
+    this.btcProvider = btcProvider;
+    this.bbnProvider = bbnProvider;
+
+    if (stakingParams.length === 0) {
+      throw new Error("No staking parameters provided");
+    }
+    this.stakingParams = stakingParams;
+  }
+
+  /**
+   * Creates an signed EOI transaction that is ready to be sent to the BBN chain.
+   * @param stakingInput - The staking inputs.
+   * @param feeRate - The fee rate in satoshis per byte.
+   * @returns The signed babylon transaction in base64 format.
+   */
+  async createEoiDelegationBbnTransaction(
+    stakingInput: StakingInputs,
+    feeRate: number,
+  ): Promise<{
+    signedBbnTx: Uint8Array;
+    stakingTx: Transaction;
+  }> {
+    const tipHeight = await this.bbnProvider.getBabylonBtcTipHeight();
+    if (tipHeight === 0) {
+      throw new Error("Babylon BTC tip height cannot be 0");
+    }
+    const inputUTXOs = await this.btcProvider.getUTXOs();
+    if (inputUTXOs.length === 0) {
+      throw new Error("No input UTXOs provided");
+    }
+    const bech32Address = await this.bbnProvider.getBbnAddress();
+    if (!bech32Address) {
+      throw new Error("Babylon address cannot be empty");
+    }
+
+    // Get the param based on the tip height
+    // EOI should always be created based on the BTC tip height from BBN chain
+    const p = getBbnParamByBtcHeight(tipHeight, this.stakingParams);
```

Comment:
Should we be getting the parameters based on the Babylon height here or based on the btc height? In the case of inconsistency between btc and and Babylon (e.g. delayed headers reporting), we might run the risk that a staking transaction utilizes the parameters from e.g. v1, while it is included in a block height that is part of a v2 block height range.